### PR TITLE
Update RMM adaptors, containers and tests to use get/set_current_device_resource_ref()

### LINKED ISSUE #1660: Update adaptors, containers and tests to use get/set_current_device_resource()
Part of #1596

### PR DESCRIPTION
**Pull request by harrism**
## Description
Closes #1660.

This adds a constructor to each MR adaptor to take a resource_ref rather than an `Upstream*`. It also updates RMM to use `get_current_device_resource_ref()` everywhere: in containers, in tests, in adaptors, Thrust allocator, polymorphic allocator, execution_policy, etc.

Importantly, this PR also modifies `set_current_device_resource()` to basically call `set_current_device_resource_ref()`. This is necessary, because while RMM C++ uses `get_current_device_resource_ref()` everywhere, the Python API still uses the raw pointer API `set_current_device_resource()`. So we need the latter to update the state for the former.  This is a temporary bootstrap to help with the refactoring.

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rmm/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by wence-** [comment]
On the Python side, we need someone to own concrete objects (we can't just provide classes that wrap `device_async_resource_ref` since that is non-owning).

Adapters also need a way to keep their upstream alive.

We are removing the `device_memory_resource` inheritance on the C++ side, but I think until we have an owning `any_resource` available, we probably need to keep something on the Cython side.

Here is a very slide-ware sketch approach for discussion:

```
cdef extern from * nogil:
    """
    struct hack {
        std::function<rmm::device_async_resource_ref> _to_ref;
        hack() = default; // unsafe, but this is all hacky anyway
        rmm::device_async_resource_ref to_ref() { return _to_ref(); }
        template<typename T>
        hack(std::unique_ptr<T> obj) : _to_ref{[&obj]() -> rmm::device_async_resource_ref { return *obj; }} {};
    };
    """
       
cdef class DeviceMemoryResource:
   cdef hack obj;
   cdef device_async_resource_ref as_ref(self):
       return self.obj.to_ref()

cdef class CudaMemoryResource(DeviceMemoryResource):
   def __cinit__(self):
       self.obj = hack(move(make_unique[cuda_memory_resource]()))
   

cdef class PoolMemoryResource(DeviceMemoryResource):
   cdef DeviceMemoryResource upstream
   def __cinit__(self, DeviceMemoryResource upstream):
      self.upstream = upstream
      self.obj = hack(move(make_unique[pool_memory_resource](upstream.to_ref())))
```

This is effectively working around not (yet) having `any_resource` available by doing a minimal type-erased owning wrapper.

**Comment by harrism** [comment]
> On the Python side, we need someone to own concrete objects (we can't just provide classes that wrap `device_async_resource_ref` since that is non-owning).

Probably should have this discussion somewhere else? (Maybe a discussion thread) I was planning for this to be a C++-only PR, replacing #1634.

File: include/rmm/mr/device/per_device_resource.hpp
Comment (abellina) [inline include/rmm/mr/device/per_device_resource.hpp:244]: cudf JNI also uses `set_per_device_resource` and `get_current_device_resource` today, so we'll also be needing to change our code to use `_ref`.

Comment (harrism) [reply include/rmm/mr/device/per_device_resource.hpp:244]: 🙏

File: include/rmm/mr/device/aligned_resource_adaptor.hpp
Comment (miscco) [inline include/rmm/mr/device/aligned_resource_adaptor.hpp:97]: Nitpick: This lambda appears 7 times, I believe it would be better if this would become an explicit function:
```cpp
template<class Upstream>
device_async_resource_ref to_device_async_resource_ref_checked(Upstream* upstream) {
    RMM_EXPECTS(nullptr != upstream, "Unexpected null upstream resource pointer.");
    return device_async_resource_ref{*upstream};
}
```

**Review by miscco — APPROVED** [review]

File: tests/mr/device/arena_mr_tests.cpp
Comment (wence-) [inline tests/mr/device/arena_mr_tests.cpp:18]:
```suggestion
#include <cuda/stream_ref>
```

And subsequent include reordering?

**Review by wence- — APPROVED** [review]
Looks good to me, with a question about the (now-redundant) class template parameter on all the adaptor MRs.

**Review by rongou — APPROVED** [review]

File: include/rmm/mr/device/aligned_resource_adaptor.hpp
Comment (harrism) [reply include/rmm/mr/device/aligned_resource_adaptor.hpp:212]: Yep.

File: include/rmm/mr/device/detail/arena.hpp
Comment (harrism) [reply include/rmm/mr/device/detail/arena.hpp:694]: clang-tidy fix.

File: tests/mr/device/arena_mr_tests.cpp
Comment (harrism) [reply tests/mr/device/arena_mr_tests.cpp:18]: Yep, blame clangd.  Good catch.

File: include/rmm/mr/device/aligned_resource_adaptor.hpp
Comment (harrism) [reply include/rmm/mr/device/aligned_resource_adaptor.hpp:91]: Right, the plan is to remove the template parameter and the `Upstream*` constructors, once we add the `resource_ref` constructors and convert all of RAPIDS to use them.  But we can't do it yet.

See https://github.com/rapidsai/rmm/issues/1457

File: include/rmm/mr/device/aligned_resource_adaptor.hpp
Comment (harrism) [reply include/rmm/mr/device/aligned_resource_adaptor.hpp:212]: Actually, I think these can be all removed except in containers and allocators that have default ctors.

File: include/rmm/mr/device/detail/arena.hpp
Comment (harrism) [reply include/rmm/mr/device/detail/arena.hpp:694]: I'm resisting the temptation to fix more clang-tidy errors in this PR. Sorry.

**Comment by harrism** [comment]
Let's not merge until I can do some more downstream testing, including JNI.

File: tests/mr/device/arena_mr_tests.cpp
Comment (harrism) [reply tests/mr/device/arena_mr_tests.cpp]: Fixed.

**Review by bdice — COMMENTED** [review]
I got about 2/3 through the changes. I’ll finish tomorrow.

File: include/rmm/mr/device/pool_memory_resource.hpp
Comment (harrism) [reply include/rmm/mr/device/pool_memory_resource.hpp:122]: Yes I updated most of these but probably missed some.

File: include/rmm/resource_ref.hpp
Comment (harrism) [reply include/rmm/resource_ref.hpp:80]: This comes from the old advice to use `if (CONSTANT == variable)` rather than `if (variable == CONSTANT)`. The reasoning was that it was common to typo `=` for `==`, leading to hard to find assignment bugs. These days we have tools to detect that, but old habits die hard.

If I follow ancient practices, maybe you *should* call me Yoda. :)

Comment (bdice) [reply include/rmm/resource_ref.hpp:80]: Ha! This is non-blocking, of course. You are the wisest Jedi so I defer to your preference.

File: tests/mr/host/mr_ref_tests.cpp
Comment (bdice) [inline tests/mr/host/mr_ref_tests.cpp:236]: Maybe this is related to the CMakeLists.txt change I noted above? What's the rationale for this change?

**Review by bdice — APPROVED** [review]
I have a couple more comments but overall I think this is a good direction and this PR is of sufficient quality that I feel comfortable approving it. Thanks for answering my questions.

File: tests/CMakeLists.txt
Comment (harrism) [reply tests/CMakeLists.txt:43]: I would like it to stay, at least until this fixes the problem: . It only applies to Debug builds. https://github.com/rapidsai/rapids-cmake/issues/634#issuecomment-2317976716

File: tests/mr/host/mr_ref_tests.cpp
Comment (harrism) [reply tests/mr/host/mr_ref_tests.cpp:236]: In debug mode, the deallocation will assert, crashing the test.

File: include/rmm/mr/device/aligned_resource_adaptor.hpp
Comment (harrism) [reply include/rmm/mr/device/aligned_resource_adaptor.hpp]: Fixed.

File: include/rmm/mr/device/aligned_resource_adaptor.hpp
Comment (harrism) [reply include/rmm/mr/device/aligned_resource_adaptor.hpp]: Removed.

File: benchmarks/device_uvector/device_uvector_bench.cu
Comment (harrism) [reply benchmarks/device_uvector/device_uvector_bench.cu]: I changed all these to not take the address.

File: include/rmm/mr/device/aligned_resource_adaptor.hpp
Comment (harrism) [reply include/rmm/mr/device/aligned_resource_adaptor.hpp:84]: Fixed.

File: include/rmm/mr/device/pool_memory_resource.hpp
Comment (harrism) [reply include/rmm/mr/device/pool_memory_resource.hpp:122]: Fixed.

File: include/rmm/resource_ref.hpp
Comment (harrism) [reply include/rmm/resource_ref.hpp:80]: In this case went with the real dinosaur code since the error string makes it obvious. :)

**Pull request merged**
